src/persist-txn/src/txns.rs

```diff
@@ -268,54 +268,41 @@ where
             .await;
         }
 
-        let mut retractions = Vec::new();
+        let mut retractions = BTreeMap::new();
         for (data_id, batch_raw, commit_ts) in self.txns_cache.unapplied_batches() {
             if ts < commit_ts {
                 break;
             }
 
             let write = self.datas.get_write(data_id).await;
-            let by_us = crate::apply_caa(write, batch_raw, commit_ts.clone()).await;
-            if by_us {
-                // NB: Protos are not guaranteed to exactly roundtrip the
-                // encoded bytes, so we intentionally use the raw batch so that
-                // it definitely retracts.
-                retractions.push((data_id, batch_raw));
-            }
-            // TODO(txn): Commit as we go? This is idempotent but that could
-            // save us duplicate work when racing.
-        }
-
-        let mut retract_ts = recent_upper(&mut self.txns_write).await;
-        loop {
-            // TODO(txn): Need to filter here to avoid double retractions.
-            // Maelstrom won't catch this bug, so leaving it in to make sure the
-            // unit tests do once they're added.
-            let retractions = retractions
-                .iter()
-                .map(|(data_id, batch_desc)| ((*data_id, *batch_desc), &retract_ts, -1))
-                .collect::<Vec<_>>();
-            if retractions.is_empty() {
-                break;
-            }
-            let res = crate::small_caa(
-                || "txns retract",
-                &mut self.txns_write,
-                &retractions,
-                retract_ts.clone(),
-                retract_ts.step_forward(),
-            )
-            .await;
-            match res {
-                Ok(()) => break,
-                Err(current) => {
-                    retract_ts = current;
-                    continue;
-                }
-            }
+            crate::apply_caa(write, batch_raw, commit_ts.clone()).await;
+            // NB: Protos are not guaranteed to exactly roundtrip the
+            // encoded bytes, so we intentionally use the raw batch so that
+            // it definitely retracts.
+            retractions.insert(batch_raw.clone(), *data_id);
```

Comment:
I think you have a comment about this somewhere else, but in case you don't, if two batches are identical then one of them will never get retracted (I think).

Reply:
Yup, there's a TODO(txn) in push_append

Aside: I looked at this for a sec recently and I _think_ we're okay in practice (there's enough in the batch to ensure it's always unique), but I want to do some more due diligence before removing the TODO